### ISSUE TITLE
Screen goes to sleep mode after playback ends

### ISSUE THREAD

**Issue by blzos** [zaps166/QMPlay2#717]
**Hello again!**

So, I play the video (locally or stream, it doesn’t matter) on the laptop, which I don’t touch (this is important) during playback, since I work on the main computer. In the power settings, it is worth dimming the screen after five minutes, turning it off after 10. During playback, after five minutes, the screen darkens **(!!!)**, video playback continues, let’s say it ended after 30 minutes. In the player settings, I have a stop after the current file. Consequently, playback stops and the screen straightaway goes into sleep mode **(!!!)**.

_Tested on Windows 7, Windows 10. Player version Qt5 and Qt6 (QMPlay2 24.04.07). Fullscreen, no fullscreen._

**It's easy to reproduce the situation:**
1. Set the Windows power settings to turn off the screen for 1 minute.
2. In a clean version of the player (portable), add a file, say 1 minute 10 seconds long, to the playlist.
3. Start playback and do not touch the mouse and keyboard

As soon as the video playback ends, the screen will go sleep mode, since more than one minute has passed since the input devices were not active. In this case, the correct behavior of the player is to start counting one minute of inactivity devices after stopping playback.

**To summarize the issue:**

Most likely the player does not send triggers to the system that she is active, however, it does not allow her to turn off the screen while playback is in progress.

For now, completely disabling screen sleep settings in Windows settings helps.

**Most suitable issues:**
https://github.com/zaps166/QMPlay2/issues/526

https://github.com/zaps166/QMPlay2/issues/359
https://github.com/zaps166/QMPlay2/issues/344

**Comment by zaps166** [comment]
Screen goes to sleep after playback ends - this is how it's working on Windows.

> In this case, the correct behavior of the player is to start counting one minute of inactivity devices after stopping playback.

I don't know how to do it so far.  I can add custom timer which will keep screen saver inhibited for a while (e.g. few seconds).

I use Windows quite rare, for sure not having screen saver there, so I don't know how people use it. Am I correct - the screen goes to sleep before next playlist entry starts playing?

**Comment by blzos** [comment]
> Am I correct - the screen goes to sleep before next playlist entry starts playing?

The screen turns off when the player finishes playing and has no more tasks. Since I'm using the option "Stop playback after evry file" then after playback ends, the player shows the logo in the playback window and the screen goes into sleep mode.

If the playback mode is "No repeating" and there are more files in the list, the player moves to the next one and starts playing it. I can assume that even in this case, if the playlist comes to an end and there were no touch  to the computer devices, the monitor will still sleep.

> I don't know how to do it so far. I can add custom timer which will keep screen saver inhibited for a while (e.g. few seconds).

As far as I remember, the system needs to be informed, you what application you have so that it considers it interactive. Presently it accepts the program not as a video player, but as, say, a notepad, where the absence of user actions is perceived as the fact that he has moved away from the computer.


**In any case, this can be solved by disabling the screen from going to sleep mode (through system settings). I hope over time there will be information from the community on how to send signals to Windows (every 59 seconds) that the user is active.
It's good that the screen doesn't turn off during playback!**

**Comment by zaps166** [comment]
I'm aware of this behavior, it's correct. QMPlay2 prevents screen sleep while playing video by rejecting Windows messages. Once video stops, QMPlay2 doesn't reject message anymore and Windows will put screen to sleep.

If monitor is **not** going to sleep when QMPlay2 changes to another video from playlist and is **not** going to sleep when QMPlay2 plays video, I can't see anything wrong in this behavior - the goal is to disallow Windows to put screen to sleep while playback is active.

Why this behavior is wrong in your opinion? Do you want to look at gradient with QMPlay2 icon after playback or...? I don't want to try to put any fake input events or do any weird workarounds.

I can at most defer it for a few seconds.

**Comment by blzos** [comment]
> Why this behavior is wrong in your opinion?

Let's just say that it's not really wrong, it's unusual, since it assumes that the user is sitting and watching the video, it ends and the screen turns off without allowing the user to do anything.

> Do you want to look at gradient with QMPlay2 icon after playback or...?

I will see that the video stream is completed and select the next one I need from the list

> I don't want to try to put any fake input events or do any weird workarounds.

If this is not done with one command in the code immediately after stopping the video, of course it is better to leave it as is for now.

**Something like that:** Private Const SPI_SETSCREENSAVETIMEOUT As Int32 = 15

https://www.codeguru.com/visual-basic/preventing-your-monitor-from-going-to-sleep-with-visual-studio-2012/

**Comment by zaps166** [comment]
Looks like `SPI_SETSCREENSAVETIMEOUT` sets screensaver timeout only (not screen sleep) and this change is permanent (it's modifying user's settings). It modifies this:
![Zrzut ekranu_20240429_100237](https://github.com/zaps166/QMPlay2/assets/6239090/55ffd49c-0399-4358-8441-793ead4bcca8)

Also `SetThreadExecutionState` doesn't prevent screensaver.

I can expand the time after playback finishes for n seconds (e.g. 5 seconds).

**Comment by zaps166** [comment]
Replaced this patch by e269e02820dea6e3c1eb8ac82419ddfc4552c896.

**Comment by blzos** [comment]
Now it works as it should! The countdown for the screen to go into sleep mode starts after the end of playback. (Windows)
